### DISCUSSION TITLE
Possibility of having ConfigBuilder mirror immutable config objects for use

### DISCUSSION THREAD

**Discussion by exteon** [symfony/symfony#43154] [Q&A]
For as long as I can remember, I have preferred using structured immutable objects to represent configuration. There are, I think, a plethora of reasons to do so:

- Clear, authoritative structure and typing; immediate static analysis detection of usage of mistyped or removed config keys
- Can marshal and persist as objects, preserving full structure and references
- Possibility of dereferencing nested object references (which are usually text keys in a text config, symfony is full of examples here)
- Some external config objects can be very costly to retrieve and should be done lazily; sometimes even some of the remote config properties are even more costly to retrieve and should be lazy-loaded. Storing them in text files makes no sense and neither projecting them to a string array just to satisfy some artificial convention.

The argument for config objects could go on, but looking at the ConfigBuilder in 5.3 I was wondering if it wouldn't just be a great idea (and half already done), if, in pair with the cached config builders, we could also have generated config classes, not in cache, but in user versioned code that could be used to represent the read config. They would be immutable, and feature a constructor with defaults taken from the TreeBuilder, as well as `fromArray` and `toArray` methods.

**Comment by stof** [comment]
Well, the `Configuration` class and its `Processor` in symfony/config are about merging configs from multiple files (which can be in different formats) and validating them. The whole layer of the config of DI extension works with arrays, and changing that would be a huge BC break for the whole ecosystem, so it's not gonna happen anytime soon (and won't ever happen if we don't find a plan to avoid making it a huge BC break).

External config objects don't make sense in the context of DI extensions. The config is necessary to build the container.

**Comment by exteon** [comment]
Well, in terms of making sense, I was thinking along two lines:
- The standalone `config` package, which does not hold notion of the DI and for which having the possibility of building config objects makes so much sense. I have seen (and written) a number of third-party services that are constructed along the lines of `__construct(Dependency $dependency, ServiceConfig $config, string $parameter)`
- In the context of the `DI`, where it could allow for instance to autowire a `ServiceConfig` instantiated from local yaml configs for dev and testing environments, but a dynamic `ServiceConfig` in production that reacts to dynamic ecosystem changes

In terms of implementation, I only envisioned the `ServiceConfig` (_parameters_ in Symfony speak) part of the service instantiation, not the `Dependency` part, which would simplify things a lot, so I envisioned just an added layer that _could_ be used if needed: so all the core process remains unchanged, but if I need to, I can use a `ServiceConfigBuilder::fromContainerBag()` that would just replace the already existing `ContainerBagInterface` option (https://symfony.com/doc/current/configuration.html#accessing-configuration-parameters) and that would internally just use `ServiceConfig::fromArray()` on the array representation in the `ContainerBag`.

So basically it's a process that I can do by hand anyways:
- create a `ServiceConfig` class
- implement static typing and validation, and `fromArray` method
- find a way of injecting this, one option would be to register this with a `Configurator`; or just explicitly build a `ServiceConfig` object in the service constructor from the `ContainerBagInterface` object

The problem is, to do this by hand takes a lot of time, it's error-prone and it's completely redundant with information I already have in the `TreeBuilder` and which is used already to build a `ConfigBuilder`, so it would seem not a great effort to have a method to also generate the pairing readable object.

As a matter of fact, this would be so useful that even in projects where I don't use Symfony at all, I could use this as a tool to generate config classes, because the TreeBuilder is such a handy tool for keeping configuration metadata together (structure, validation, documentation, defaults etc).

So all in all, my scope when I wrote this was much narrower and I see a much simpler solution, unless I'm missing things, which I probably am...

**Reply by stof** [reply]
The semantic configuration processed by DI extension is **not** the same concept that Symfony parameters. More than half of the configuration settings of bundles are about changing the way services are wired together, not about setting parameters.
All this happens during building the container, not when using the ContainerBagInterface

**Reply by stof** [reply]
and note that it is fine that you use the TreeBuilder of symfony/config in a standalone way. But the main consumer of that component in the Symfony ecosystem is the system of DI extensions allowing bundles to configure the container (which is the reason why this feature was added in symfony/config in the first place). So we won't rewrite that feature of symfony/config in a way making it harder to use for that usage.

**Reply by exteon** [reply]
Ok, so I'm a bit lost here, because there's so much about Symfony that I'm learning just now. If you have the patience to bear with me for a moment, I'd really need your help to understand this, I think I jumped in the middle of what seems to be a bigger problem. So let's take a concrete problem: 

I have an out-of-the-box service that looks like:

```php
class Service {
    public function __construct(Depndency $dependency, ServiceConfig $config){
    }
}

class ServiceConfig {
    ...
    public function getParam(): string;
    
    /** @return TransportConfig[] */
    public function getTransportConfigs(): array; 

    public static function fromArray(array $array): self;
}
```
This is already implemented and the config was loaded via `::fromArray()` from a yaml file:
```yaml
param: value
transport-configs:
  - channel: value
  - channel: value
```

What I would need to understand is how do I bring this inside a Symfony container so that: `Dependency` is autowired and I also get the ServiceConfig object? I'm not concerned about where the content for the config needs to come from: either `services.yaml` or a separate config file is just as fine. What's the cleanest way to achieve this?
